streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -648,55 +648,62 @@ sealed trait ZChannel[-Env, -InErr, -InElem, -InDone, +OutErr, +OutElem, +OutDon
       for {
         input       <- SingleProducerAsyncInput.make[InErr, InElem, InDone]
         queueReader  = ZChannel.fromInput(input)
-        outgoing    <- Queue.bounded[ZIO[Env1, Unit, Either[OutDone, OutElem2]]](bufferSize)
+        outgoing    <- Queue.bounded[Fiber[Either[Unit, OutDone], OutElem2]](bufferSize)
         _           <- scope.addFinalizer(outgoing.shutdown)
-        errorSignal <- Promise.make[Unit, Nothing]
+        errorSignal <- Promise.make[Nothing, Unit]
         permits     <- Semaphore.make(n.toLong)
         failure     <- Ref.make[Cause[OutErr1]](Cause.empty)
-        pull        <- (queueReader >>> self).toPullIn(scope)
-        _ <- pull
-               .foldCauseZIO(
-                 cause =>
-                   failure.update(_ && cause) *> outgoing.offer(Exit.failCause(Cause.unit)) *> Exit.failCause(
-                     Cause.unit
-                   ),
-                 {
-                   case Left(outDone) =>
-                     permits.withPermits(n.toLong)(ZIO.unit).interruptible *> outgoing.offer(ZIO.succeed(Left(outDone)))
-                   case Right(outElem) =>
-                     for {
-                       p     <- Promise.make[Unit, OutElem2]
-                       latch <- Promise.make[Nothing, Unit]
-                       _     <- outgoing.offer(p.await.map(Right(_)))
-                       _ <- permits.withPermit {
-                              latch.succeed(()) *>
-                                ZIO.uninterruptibleMask { restore =>
-                                  restore(errorSignal.await).raceFirstAwait(
-                                    restore(f(outElem))
-                                      .catchAllCause(cause => failure.update(_ && cause) *> Exit.failCause(Cause.unit))
-                                  )
-                                }.foldCauseZIO(
-                                  _ => p.refailCause(Cause.unit) *> errorSignal.refailCause(Cause.unit),
-                                  p.succeed
-                                )
-                            }.forkIn(scope)
-                       _ <- latch.await
-                     } yield ()
+        pull        <- (queueReader >>> self).toPullInAlt(scope)
+        _ <- ZIO.fiberIdWith { fiberId =>
+               pull.flatMap { outElem =>
+                 val latch = Promise.unsafe.make[Nothing, Unit](fiberId)(Unsafe.unsafe)
+                 for {
+                   f <- permits
+                          .withPermit(
+                            latch.succeed(()) *> f(outElem)
+                              .catchAllCause(cause =>
+                                failure.update(_ && cause).unless(cause.isInterrupted) *>
+                                  errorSignal.succeed(()) *>
+                                  ZChannel.failLeftUnit
+                              )
+                          )
+                          .interruptible
```

Comment:
There are a few small discrepancies in the behaviour here due to `interruptible` from previously that I don't know how much of an actual issue they might be but I'll bring them up to see what you think:

1. If the stream is run in an uninterruptible region, now it can be interrupted at pretty much any point as the `interruptible` will also make the effect within `withPermit` interruptible.
2. If `f(outElem)` was interruptible and the stream was interrupted, previously the `catchAllCause` would be guaranteed to run. Now since `catchAllCause` is within an interruptible region it won't be called in this case

What do you think? Might this be an actual issue?

Reply:
this is very confusing but no, my reading is that adding this `interruptible` only restores the behaviour that I want, i.e. undo any potential damage caused by someone writing `stream.uninterruptible` (which is still crazy to me but anyway), it doesn't change the fact that `withPermit` will acquire the semaphore and restore it if acquired no matter what—the section inside is not run uninterruptibly.